### PR TITLE
feat(patching): support pool patching on XS 7.1

### PR DESCRIPTION
**Pull request by pdonias**
See vatesfr/xo-web#2058

### CODE REVIEW

File: src/xapi/utils.js [hunk 1]
```diff
@@ -383,3 +383,6 @@ export const useUpdateSystem = host => {
   // Match Xen Center's condition: https://github.com/xenserver/xenadmin/blob/f3a64fc54bbff239ca6f285406d9034f57537d64/XenModel/Utils/Helpers.cs#L420
   return versionSatisfies(host.software_version.platform_version, '^2.1.1')
 }
+
+export const isSrAdequate = (sr, minSize) =>
```
Comment (julien-f) [inline src/xapi/utils.js]: *Adequate* for what?

File: src/xapi/index.js [hunk 2]
```diff
@@ -2192,21 +2189,25 @@ export default class Xapi extends XapiBase {
...
+      throw new Error('no SR available')
     }
 
-    return this._createTemporaryVdiOnSr(stream, sr, name_label, name_description)
+    return this.createTemporaryVdiOnSr(stream, sr, name_label, name_description)
+  }
+
+  async findAvailableSharedSr (minSize) {
+    return find(
+      filter(this.objects.all, { $type: 'sr', shared: true }),
```
Comment (julien-f) [inline src/xapi/index.js]: Do not filter: integrate the test in `find`'s predicate.

File: src/xapi/index.js [hunk 3]
```diff
@@ -2192,21 +2189,25 @@ export default class Xapi extends XapiBase {
...
   async createTemporaryVdiOnHost (stream, hostId, name_label, name_description) {
-    let sr = this.pool.$default_SR
-
-    if (!sr || sr.physical_size - sr.physical_utilisation < stream.length) {
-      sr = find(
-        mapToArray(this.getObject(hostId).$PBDs, '$SR'),
-        sr => sr && sr.content_type === 'user' && sr.physical_size - sr.physical_utilisation >= stream.length
-      )
+    const sr = find(
+      mapToArray(this.getObject(hostId).$PBDs, '$SR'),
```
Comment (julien-f) [inline src/xapi/index.js]: Avoid the `mapToArray`.

File: src/xapi/mixins/patching.js [hunk 4]
```diff
@@ -202,6 +205,13 @@ export default {
     return installable
   },
 
+  _isPoolPatchInstallableOnPool (patchUuid) {
+    return every(
+      filter(this.objects.all, { $type: 'host' }),
```
Comment (julien-f) [inline src/xapi/mixins/patching.js]: Should be pretty easy to avoid this `filter`.

File: src/xapi/mixins/patching.js [hunk 5]
```diff
@@ -288,59 +281,113 @@ export default {
...
 
     await this.call('pool_patch.apply', patch.$ref, host.$ref)
   },
 
-  _installPatchUpdateOnHost (patchUuid, host) {
-    return Promise.all([ this._downloadPatchAndInstall(patchUuid, host.$id), this._ejectToolsIsos(host.$ref) ])
-  },
+  // patform_version >= 2.1.1
+  _installPatchUpdateOnHost: deferrable(async function ($defer, patchUuid, host) {
+    const [ vdi ] = await Promise.all([ this._getUpdateVdi($defer, patchUuid, host.$id), this._ejectToolsIsos(host.$ref) ])
```
Comment (julien-f) [inline src/xapi/mixins/patching.js]: Add new lines.

File: src/xapi/mixins/patching.js [hunk 6]
```diff
@@ -288,59 +281,113 @@ export default {
...
   },
 
-  async installPoolPatchOnHost (patchUuid, hostId) {
-    debug('installing patch %s', patchUuid)
+  // -----------------------------------------------------------------
 
-    return this._checkSoftwareVersionAndInstallPatch(patchUuid, hostId)
+  // platform_version < 2.1.1
+  async _installPoolPatchOnAllHosts (patchUuid) {
+    const [ patch ] = await Promise.all([ this._getOrUploadPoolPatch(patchUuid), this._ejectToolsIsos() ])
```
Comment (julien-f) [inline src/xapi/mixins/patching.js]: New lines.

File: src/xapi/mixins/patching.js [hunk 7]
```diff
@@ -288,59 +281,113 @@ export default {
...
+
+    await this.call('pool_patch.pool_apply', patch.$ref)
   },
 
-  // -----------------------------------------------------------------
+  // platform_version >= 2.1.1
+  _installPatchUpdateOnAllHosts: deferrable(async function ($defer, patchUuid) {
+    let [ vdi ] = await Promise.all([ this._getUpdateVdi($defer, patchUuid), this._ejectToolsIsos() ])
```
Comment (julien-f) [inline src/xapi/mixins/patching.js]: New lines.

File: src/xapi/mixins/patching.js [hunk 8]
```diff
@@ -370,11 +417,32 @@ export default {
...
-      host => this.installAllPoolPatchesOnHost(host.$id)
-    ))
+    const installableByUuid = assign(
+      {},
+      ...await Promise.all(mapFilter(this.objects.all, host => {
+        if (host.$type !== 'host') {
+          return
+        }
+
+        return this._listMissingPoolPatchesOnHost(host)
```
Comment (julien-f) [inline src/xapi/mixins/patching.js]: Reverse the condition and move this in the  `if`.

File: src/xapi/utils.js (hunk 1 above)
Comment (pdonias) [inline src/xapi/utils.js]: Is the SR `sr` available to host a new VDI of size `minSize`

Comment (julien-f) [inline src/xapi/utils.js]: Maybe the name could reflect this more: `canSrHaveNewVdiOfSize` or something similar.

File: src/xapi/utils.js [hunk 9]
```diff
@@ -383,3 +383,6 @@ export const useUpdateSystem = host => {
...
+  sr && sr.content_type === 'user' && sr.physical_size - sr.physical_utilisation >= minSize
```
Comment (julien-f) [inline src/xapi/utils.js]: `sr != null`

**Comment by julien-f** [comment]
Can you rebase this please :)

**Pull request merged**